File: app/src/main/java/com/example/cheng/myapplication/TLinearLayout.java
package com.example.cheng.myapplication;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

/**
 * Created by biao.cheng on 2017/10/20.
 */

public class TLinearLayout extends LinearLayout {

    private int childNum = 3;

    public void setChildNum(int childNum) {
        this.childNum = childNum;
    }

    private int w;

    public TLinearLayout(Context context) {
        super(context);
        init();
    }

    public TLinearLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public TLinearLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        this.setOrientation(LinearLayout.HORIZONTAL);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        this.w = w;
        Log.i("aaa", "onSizeChanged=" + getMeasuredWidth() + ",w=" + w);
//        setItemNum(4);
    }

    public void setItemNum(int num) {
        this.removeAllViews();
        Log.i("aaaa", "addView" + num);
        for (int i = 0; i < num; i++) {
            TextView tv = new TextView(getContext());
            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(getwidth() / num, ViewGroup.LayoutParams.MATCH_PARENT);
            tv.setText("num" + i);
            tv.setTextColor(Color.BLUE);
            tv.setGravity(Gravity.CENTER);
            this.addView(tv, lp);
        }
    }

    private int getwidth() {
        return w;
    }

}

File: app/src/main/java/com/example/cheng/myapplication/TLinearLayout.kt
package com.example.cheng.myapplication

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

/**
 * Created by biao.cheng on 2017/10/20.
 */

class TLinearLayout : LinearLayout {

    private var childNum = 3

    private var w: Int = 0

    fun setChildNum(childNum: Int) {
        this.childNum = childNum
    }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        this.orientation = LinearLayout.HORIZONTAL
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        this.w = w
        Log.i("aaa", "onSizeChanged=$measuredWidth,w=$w")
        //        setItemNum(4);
    }

    fun setItemNum(num: Int) {
        this.removeAllViews()
        Log.i("aaaa", "addView" + num)
        for (i in 0 until num) {
            val tv = TextView(context)
            val lp = LinearLayout.LayoutParams(getwidth() / num, ViewGroup.LayoutParams.MATCH_PARENT)
            tv.text = "num" + i
            tv.setTextColor(Color.BLUE)
            tv.gravity = Gravity.CENTER
            this.addView(tv, lp)
        }
    }

    private fun getwidth(): Int {
        return w
    }

}
